src/solver/crout/crout.hpp

```diff
@@ -0,0 +1,147 @@
+/*************************************************************************
+ * Copyright (C) 2018-2022 Blue Brain Project
+ *
+ * This file is part of NMODL distributed under the terms of the GNU
+ * Lesser General Public License. See top-level LICENSE file for details.
+ *************************************************************************/
+#pragma once
+
+/**
+ * \dir
+ * \brief Solver for a system of linear equations : Crout matrix decomposition
+ *
+ * \file
+ * \brief Implementation of Crout matrix decomposition (LU decomposition) followed by
+ * Forward/Backward substitution
+ */
+
+#include <Eigen/Core>
+#include <cmath>
+
+#if defined(CORENEURON_ENABLE_GPU) && !defined(DISABLE_OPENACC)
+#include "coreneuron/utils/offload.hpp"
+#endif
+
+namespace nmodl {
+namespace crout {
+
+/**
+ * \brief Crout matrix decomposition : in-place LU Decomposition of matrix A.
+ *
+ * LU decomposition function.
+ * Implementation details : (Legacy code) coreneuron/sim/scopmath/crout_thread.cpp
+ *
+ * \param n The number of rows or columns of the matrix A
+ * \param A matrix of size nxn : in-place LU decomposition (C-style arrays : row-major order)
+ * \param pivot matrix of size n : The i-th element is the pivot row interchanged with row i
+ */
+#if defined(CORENEURON_ENABLE_GPU) && !defined(DISABLE_OPENACC)
+nrn_pragma_acc(routine seq)
+nrn_pragma_omp(declare target)
+#endif
+template <typename T>
+EIGEN_DEVICE_FUNC inline void Crout(int n, T* A, int* pivot) {
+    int i, j, k;
+    T *p_k, *p_row, *p_col;
+    T max;
```

Comment:
can I be `nitpicky`? : declaring these variables at local scope would be nice? (e.g. when I see `j, k` being at the top level, I was wondering whether their values are used across internal for loops)

Reply:
I just wanted to be as close as possible to the legacy code. Therefore, I would prefer to leave it as is ;)